spack find external support for OpenCV and OpenBLAS

bvanessen opened the pull request:
Added support for finding the OpenCV package via the find external
command.  Included support for identifying variants based on available
shared libraries.

Added support to finding the OpenBLAS package via the find external
command.

Enabled pacakges to show that they can be discovered via the find
external command in the info message.

bvanessen committed:
Added support for finding the OpenCV package via the find external
command.  Included support for identifying variants based on available
shared libraries.

Added support to finding the OpenBLAS package via the find external
command.

Enabled pacakges to show that they can be discovered via the find
external command in the info message.

bvanessen committed:
Removed debugging statemenets

bvanessen commented:
@alalazo If you get a chance to look over this PR tonight, that would be great.

alalazo reviewed (approved):
LGTM. Code for finding executables and libraries is very similar, if not equal, in most packages. We should probably do something to gather it together in a core function that can be reused, but that's for a later PR.

alalazo commented on var/spack/repos/builtin/packages/opencv/package.py:
```diff
@@ -233,10 +234,16 @@ class Opencv(CMakePackage, CudaPackage):
     for mod in modules:
         # At least one of these modules must be enabled to build OpenCV
         variant(mod, default=False, description="Include opencv_{0} module".format(mod))
+        lib = f'libopencv_{mod}'
```
Can't use f-strings, we need to support Python 2.7

bvanessen replied:
Okay, I just pushed a fix.

alalazo reviewed (changes requested):
I'll merge as soon as tests pass :slightly_smiling_face:

adamjstewart commented:
Can this be split into 2 separate PRs? These 2 packages are independent of each other.

bvanessen committed:
Fixed flake8 and Python 2.7 compatibility

bvanessen committed:
Fixed flake8

alalazo commented:
> Can this be split into 2 separate PRs? These 2 packages are independent of each other.

I think that's a good advice in general, but let's not add a burden on the submitter for this time :slightly_smiling_face:

adamjstewart reviewed (changes requested):
These won't work on macOS which uses `.dylib` instead of `.so`. Here are some example file names:
```console
$ spack find -pv opencv
opencv@4.5.4~1394~ade~alphamat~android_mediandk~android_native_camera~annotation~aruco~avfoundation~barcode~bgsegm~bioinspired~calib3d~cap_ios~carotene~ccalib~clp~cpufeatures~createsamples~cublas~cuda~cudaarithm~cudabgsegm~cudacodec~cudafeatures2d~cudafilters~cudaimgproc~cudalegacy~cudaobjdetect~cudaoptflow~cudastereo~cudawarping~cudev~cudnn~cufft~cvv~datasets~directx~dnn~dnn_objdetect~dnn_superres~dpm~dshow~eigen~face~fast-math~features2d~ffmpeg~flann~freetype~fuzzy~gapi~gdal~gtk~hdf~hfs~highgui~hpx~img_hash~imgcodec_hdr~imgcodec_pfm~imgcodec_pxm~imgcodec_sunraster+imgcodecs+imgproc~intensity_transform~interactive-calibration~ipo~ipp~itt~jasper~java~java_bindings_generator+jpeg~lapack~line_descriptor~matlab~mcc~ml~model-diagnostics~msmf~msmf_dxva~nonfree~objc~objc_bindings_generator~objdetect~onnx~opencl~opencl_d3d11_nv~openexr~opengl~openjpeg~openmp~optflow~phase_unwrapping~photo~plaidml~plot+png~powerpc~protobuf~pthreads_pf~python2+python3+python_bindings_generator~python_tests~qt~quality~quirc~rapid~reg~rgbd~saliency~sfm~shape+shared~stereo~stitching~structured_light~superres~surface_matching~tbb~tengine~tesseract~text+tiff~tracking~traincascade~ts~v4l~version~video~videoio~videostab~visualisation~viz~vtk~vulcan~webp~wechat_qrcode~win32ui~world~xfeatures2d~ximgproc~xobjdetect~xphoto build_type=RelWithDebInfo patches=1d30f12,2aea289,3cab451  /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/opencv-4.5.4-a6ggmspsgsc6cvelpnfqufec5c6s76xt
$ ls /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/opencv-4.5.4-a6ggmspsgsc6cvelpnfqufec5c6s76xt/lib/
cmake				libopencv_core.dylib		libopencv_imgcodecs.dylib	libopencv_imgproc.dylib
libopencv_core.4.5.4.dylib	libopencv_imgcodecs.4.5.4.dylib	libopencv_imgproc.4.5.4.dylib	python3.9
libopencv_core.4.5.dylib	libopencv_imgcodecs.4.5.dylib	libopencv_imgproc.4.5.dylib
```

bvanessen committed:
Added a helper function in the filesystem utilities to get a list of
the legal library extensions.

Updated the OpenCV and OpenBLAS packages to use the extensible search
mechanism for library extensions on multiple OS platforms.

bvanessen committed:
Corrected how find externals works on Darwin for OpenCV and OpenBLAS
to accommodate that the version numbers are placed before the file
extension instead of after it, as on Linux.

Also fixed error introduced by using the lib_extensions function name.

bvanessen committed:
Changed the function name for the library extensions

bvanessen committed:
Simplified and corrected bug in pattern detection in OpenCV.  Fixed Flake8.

bvanessen committed:
Simplified the logic for OpenBLAS and fixed Flake8.

bvanessen committed:
Removed unnecessary code

bvanessen commented:
> These won't work on macOS which uses `.dylib` instead of `.so`. Here are some example file names:
> 
> ```
> $ spack find -pv opencv
> opencv@4.5.4~1394~ade~alphamat~android_mediandk~android_native_camera~annotation~aruco~avfoundation~barcode~bgsegm~bioinspired~calib3d~cap_ios~carotene~ccalib~clp~cpufeatures~createsamples~cublas~cuda~cudaarithm~cudabgsegm~cudacodec~cudafeatures2d~cudafilters~cudaimgproc~cudalegacy~cudaobjdetect~cudaoptflow~cudastereo~cudawarping~cudev~cudnn~cufft~cvv~datasets~directx~dnn~dnn_objdetect~dnn_superres~dpm~dshow~eigen~face~fast-math~features2d~ffmpeg~flann~freetype~fuzzy~gapi~gdal~gtk~hdf~hfs~highgui~hpx~img_hash~imgcodec_hdr~imgcodec_pfm~imgcodec_pxm~imgcodec_sunraster+imgcodecs+imgproc~intensity_transform~interactive-calibration~ipo~ipp~itt~jasper~java~java_bindings_generator+jpeg~lapack~line_descriptor~matlab~mcc~ml~model-diagnostics~msmf~msmf_dxva~nonfree~objc~objc_bindings_generator~objdetect~onnx~opencl~opencl_d3d11_nv~openexr~opengl~openjpeg~openmp~optflow~phase_unwrapping~photo~plaidml~plot+png~powerpc~protobuf~pthreads_pf~python2+python3+python_bindings_generator~python_tests~qt~quality~quirc~rapid~reg~rgbd~saliency~sfm~shape+shared~stereo~stitching~structured_light~superres~surface_matching~tbb~tengine~tesseract~text+tiff~tracking~traincascade~ts~v4l~version~video~videoio~videostab~visualisation~viz~vtk~vulcan~webp~wechat_qrcode~win32ui~world~xfeatures2d~ximgproc~xobjdetect~xphoto build_type=RelWithDebInfo patches=1d30f12,2aea289,3cab451  /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/opencv-4.5.4-a6ggmspsgsc6cvelpnfqufec5c6s76xt
> $ ls /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/opencv-4.5.4-a6ggmspsgsc6cvelpnfqufec5c6s76xt/lib/
> cmake				libopencv_core.dylib		libopencv_imgcodecs.dylib	libopencv_imgproc.dylib
> libopencv_core.4.5.4.dylib	libopencv_imgcodecs.4.5.4.dylib	libopencv_imgproc.4.5.4.dylib	python3.9
> libopencv_core.4.5.dylib	libopencv_imgcodecs.4.5.dylib	libopencv_imgproc.4.5.dylib
> ```

@adamjstewart Okay I have added support for finding OS X based dynamic libraries.

bvanessen commented:
@tgamblin @becker33 @scheibelp Since this now touches a function in the main part of Spack, I didn't know if you would want to look at it.

bvanessen commented:
@adamjstewart or @alalazo Is there anything more on this PR that you want to see?

adamjstewart commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -109,12 +110,17 @@ def path_contains_subdirectory(path, root):
     return norm_path.startswith(norm_root)
 
 
+def library_extensions():
```
This should prob be a variable instead of a function

bvanessen replied:
Done

adamjstewart commented on lib/spack/spack/detection/path.py:
```diff
@@ -87,7 +87,9 @@ def libraries_in_ld_library_path(path_hints=None):
-    path_hints = path_hints or spack.util.environment.get_path('LD_LIBRARY_PATH')
+    path_hints = path_hints or \
+        spack.util.environment.get_path('LD_LIBRARY_PATH') or \
+        spack.util.environment.get_path('DYLD_LIBRARY_PATH')
```
Why do we only search `LD_LIBRARY_PATH` and not `LIBRARY_PATH`? Also, you should add `DYLD_FALLBACK_LIBRARY_PATH`, this is what Spack uses.

bvanessen replied:
Done.

adamjstewart reviewed (commented):
Still need to actually test the code but here is a first pass review

bvanessen committed:
Addressing reviewer suggestions.

adamjstewart commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -109,12 +110,16 @@ def path_contains_subdirectory(path, root):
     return norm_path.startswith(norm_root)
 
 
+"""This generates the library filenames that may appear on any OS.
+"""
```
```suggestion
#: This generates the library filenames that may appear on any OS.
```

bvanessen committed:
Update lib/spack/llnl/util/filesystem.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

bvanessen commented:
@alalazo @adamjstewart I just rebased the PR to make it easier to merge.

adamjstewart commented on var/spack/repos/builtin/packages/openblas/package.py:
```diff
@@ -6,6 +6,8 @@
 import os
 import re
 
+from llnl.util import filesystem
```
I would just use `from llnl.util.filesystem import library_extensions`

bvanessen replied:
Done

adamjstewart commented:
Tried `spack load opencv` and `spack external find opencv` but it didn't find the installation even though it's in my `DYLD_FALLBACK_LIBRARY_PATH`. Do you have access to a macOS machine to test this? If not I can take a closer look and try to figure out why it's not working.

bvanessen commented:
> Tried `spack load opencv` and `spack external find opencv` but it didn't find the installation even though it's in my `DYLD_FALLBACK_LIBRARY_PATH`. Do you have access to a macOS machine to test this? If not I can take a closer look and try to figure out why it's not working.

@adamjstewart This is what worked for me on OS X `LD_LIBRARY_PATH=/usr/local/Cellar/openblas/0.3.20/ spack -d external find openblas`

bvanessen committed:
Addressing reviewer comment

adamjstewart commented on var/spack/repos/builtin/packages/openblas/package.py:
```diff
@@ -145,6 +149,17 @@ class Openblas(MakefilePackage):
+    @classmethod
+    def determine_version(cls, lib):
+        ver = None
+        lib_extensions = library_extensions
```
I don't think this variable is needed, we can just use `library_extensions` directly.

bvanessen replied:
Done

bvanessen committed:
Fixed bug in that all of the system library paths were not being searched, only the first defined one.

bvanessen commented:
> I also tried setting `LD_LIBRARY_PATH` to the same value as `DYLD_FALLBACK_LIBRARY_PATH` but `spack external find opencv` still doesn't pick it up. Can you try installing OpenCV with Spack and see if `spack load opencv && spack external find opencv` works for you? Maybe Homebrew uses a different installation layout?

Can you try again, I think that I have fixed this.

adamjstewart commented:
Thanks, that seems to have fixed it!

Another possible bug I've found. When there are multiple copies of OpenCV in `*LIBRARY_PATH`, Spack only adds a single one to my `packages.yaml`. This is quite different than how `spack external find` works for `PATH` when searching for binaries. Is there some logic that leaves early once one library is found?

Also, I noticed that the variant detection adds `+foo` for all libraries that are found, but doesn't add `~bar` for libraries that aren't found. I think it would be helpful to add `~bar` too.

bvanessen commented:
> Thanks, that seems to have fixed it!
> 
> Another possible bug I've found. When there are multiple copies of OpenCV in `*LIBRARY_PATH`, Spack only adds a single one to my `packages.yaml`. This is quite different than how `spack external find` works for `PATH` when searching for binaries. Is there some logic that leaves early once one library is found?
> 
> Also, I noticed that the variant detection adds `+foo` for all libraries that are found, but doesn't add `~bar` for libraries that aren't found. I think it would be helpful to add `~bar` too.

@adamjstewart I don’t know why it only finds one.  I didn’t do anything specifically to stop after the first.  I can check.  I didn’t know if we would want to constrain it with negatives on the features if they weren’t found, but I guess that I could.

bvanessen commented:
@alalazo Do you have any idea of why only one library would be returned?

alalazo commented on var/spack/repos/builtin/packages/opencv/package.py:
```diff
@@ -873,6 +883,43 @@ class Opencv(CMakePackage, CudaPackage):
+                match = pattern.search(lib)
+                if match and not match.group(2) == 'core':
+                    variants.append('+' + match.group(2))
+        results.append(' '.join(variants))
+        return results
```
Shouldn't this be:
```suggestion
        return ' '.join(variants)
```
?

alalazo commented on var/spack/repos/builtin/packages/opencv/package.py:
```diff
@@ -873,6 +883,43 @@ class Opencv(CMakePackage, CudaPackage):
+
+    @classmethod
+    def determine_variants(cls, libs, version_str):
+        results = []
```
```suggestion
```

bvanessen committed:
Apply suggestions from code review

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

bvanessen committed:
Explicitly disable variants for OpenCV libraries that are not detected.

bvanessen commented:
@adamjstewart I added the support to mark the variants not used.

adamjstewart reviewed (approved):
Confirmed that this works on macOS. It picks up my OpenCV installation and correctly determines the version and both positive and negative variants.

However, it's still not picking up multiple installations when they are all in my `*LIBRARY_PATH`. But that was already broken, so I'm fine with merging this as is and fixing that in a separate PR if you want.
